### PR TITLE
Homework4: Anton Stychnevsky

### PR DESCRIPTION
**Pull request by Stychnevsky**
Homework # 4

### CODE REVIEW

File: HW4 - task1.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+def func_calls_history(func):
+    calls_history = []  # don't use nonlocal in wrapper because list is mutable
+
+    def wrapper(*args, **kwargs):
+        func_call_result = func(*args, **kwargs)
+        calls_history.append({'func': func.__name__,
+                              'end_time': strftime("%Y-%m-%d %T", gmtime()),
+                              'result': func_call_result
+                              })
```
Comment (Stychnevsky) [inline HW4 - task1.py:45]: There is frequent situation when we have some function\input which return values(more than one and similar in every call), and this function\input is called unknown number of times (like "calls_history" in my program). What is the best type to save such sequences? I used list of dicts, but may be list of lists or list of tuples is better? Or may be it's better to create a class?

File: HW4 - task2.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ get_ranges([4,7,10]) // "4,7,10"
+ get_ranges([2, 3, 8, 9]) // "2-3,8-9"
+"""
+
+
+def get_ranges(lst):
+    if len(lst) == 1:
+        return str(lst[0])
+
+    new_lst = ''
```
Comment (belovmd) [inline HW4 - task2.py]: why variable with name list keeps string?

File: HW4 - task1.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+from time import strftime
+
+
+def func_calls_history_v1(func):
+    """decorator which save history for all functions
+
+    There is decision for first version of task - decorator save all functions
+    called during program running
+    """
+    if not hasattr(func_calls_history_v1, 'calls_history'):
```
Comment (belovmd) [inline HW4 - task1.py:17]: good solution

**Review by belovmd — COMMENTED** [review]
please address these minor issues

File: HW4 - task1.py (hunk 1 above)
Comment (belovmd) [reply HW4 - task1.py:45]: you solution is OK

**Review by belovmd — COMMENTED** [review]
9

**Pull request closed without merging**